File: Problema3/challenge.java
/**
 * @file challenge.java
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-04
 * Compile with: javac Problema3/challenge.java
 * Run with: java Problema3.challenge
 */
package Problema3;

// la firma vera: 4v3v1_v0gl14 [di reversare ...] (l'ho accorciata per comodita')
public class challenge {
    public static boolean checkSignature(String signature) {
        if(signature.length() != 12)    // la firma vera e' lunga 12 caratteri
            return false;

        // grazie copilot per questo avermi tradotto il codice C -> Java
        if(signature.codePointAt(0) * signature.codePointAt(0) - 242 * signature.codePointAt(0) != -9880)
            return false;
        if(signature.codePointAt(1) * signature.codePointAt(1) + 3 * signature.codePointAt(1) != 14278)
            return false;
        if(signature.codePointAt(2) * signature.codePointAt(2) - 54 * signature.codePointAt(2) != -153)
            return false;
        if(signature.codePointAt(3) * signature.codePointAt(3) - 61 * signature.codePointAt(3) != 6726)
            return false;
        if(signature.codePointAt(4) * signature.codePointAt(4) - 2 * signature.codePointAt(4) != 2303)
            return false;
        if(signature.codePointAt(5) * signature.codePointAt(5) + 101 * signature.codePointAt(5) != 18620)
            return false;
        if(signature.codePointAt(6) * signature.codePointAt(6) + 90 * signature.codePointAt(6) != 24544)
            return false;
        if(signature.codePointAt(7) * signature.codePointAt(7) - 72 * signature.codePointAt(7) != -1152)
            return false;
        if(signature.codePointAt(8) * signature.codePointAt(8) + 25 * signature.codePointAt(8) != 13184)
            return false;
        if(signature.codePointAt(9) * signature.codePointAt(9) - 260 * signature.codePointAt(9) != -16416)
            return false;
        if(signature.codePointAt(10) * signature.codePointAt(10) - 526 * signature.codePointAt(10) != -23373)
            return false;
        if(signature.codePointAt(11) * signature.codePointAt(11) + 425 * signature.codePointAt(11) != 24804)
            return false;
        return true;    // se non e' stato trovato nessun errore, la firma e' corretta
    }
    public static void main(String[] args) {
        if(args.length != 1) { // se non e' stato passato un argomento
            System.out.println("Usage: java challenge <signature>");
            return;
        }
        String buffer = args[0]; // prendo la firma passata come argomento
        
        if(checkSignature(buffer)) // controllo se la firma e' corretta
            System.out.println("Congrats! You have found the secret code!"); // se e' corretta, stampo il codice segreto
        else
            System.out.println("Sorry, the secret code is wrong!"); // altrimenti stampo che la firma e' sbagliata
    }
}

File: Problema3/challenge.kt
/**
 * @file challenge.kt
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-04
 * Compile with: kotlinc challenge.kt -include-runtime -d challenge.jar
 * Run with: java -jar challenge.jar
 */
import kotlin.io.*

// la firma vera: 4v3v1_v0gl14 [di reversare ...] (l'ho accorciata per comodita')
fun checkSignature(signature: String): Boolean {    // funzione che controlla la firma
    if(signature.length != 12)    // la firma vera e' lunga 12 caratteri
        return false;

    if(signature[0].code * signature[0].code - 242 * signature[0].code != -9880)
        return false;
    if(signature[1].code * signature[1].code + 3 * signature[1].code != 14278)
        return false;
    if(signature[2].code * signature[2].code - 54 * signature[2].code != -153)
        return false;
    if(signature[3].code * signature[3].code - 61 * signature[3].code != 6726)
        return false;
    if(signature[4].code * signature[4].code - 2 * signature[4].code != 2303)
        return false;
    if(signature[5].code * signature[5].code + 101 * signature[5].code != 18620)
        return false;
    if(signature[6].code * signature[6].code + 90 * signature[6].code != 24544)
        return false;
    if(signature[7].code * signature[7].code - 72 * signature[7].code != -1152)
        return false;
    if(signature[8].code * signature[8].code + 25 * signature[8].code != 13184)
        return false;
    if(signature[9].code * signature[9].code - 260 * signature[9].code != -16416)
        return false;
    if(signature[10].code * signature[10].code - 526 * signature[10].code != -23373)
        return false;
    if(signature[11].code * signature[11].code + 425 * signature[11].code != 24804)
        return false;
    return true;    // se non e' stato trovato nessun errore, la firma e' corretta
}

fun main(){
    print("Enter the secret code: ")        // chiede la firma
    val signature: String = readLine()!!    // legge la firma
    if(checkSignature(signature)) {         // controlla la firma
        println("Congrats! You have found the secret code!")    // se è corretta, stampa il messaggio
    } else {
        println("Sorry, the secret code is wrong!")             // altrimenti stampa il messaggio
    }
}